cordapp-upgrades/README.md

```diff
@@ -0,0 +1,158 @@
+Corda 3.
+
+Compiled against version 1 of the contracts CorDapp.
+
+### Version 2
+This version of the CorDapp upgrades to use the new version of the FinalityFlow present in Corda 4. To do this such that
+ the app remains compatible with Version 1, the following must be done:
+ - The flows using the new FinalityFlow must be versioned
+ - The flows must detect if the counterparty is using the old version, and revert to the old API accordingly
+   
+Compiled against version 1 of the contracts CorDapp.
+
+### Version 4
+This version adds the flows required to default on an Obligation.
+
+Compiled against version 2 of the contracts CorDapp.
+
+# Upgrade scenarios
+
+This section described the upgrade scenarios that can be explored using the workflows and contracts in this CorDapp.
+Note that all commands should be run from the root of the CorDapp directory structure (i.e. the location of this Readme).
+
+## Scenario 1: Finality Flow Upgrade
+
+This scenario covers upgrading an app to run on V4 by upgrading its usage of FinalityFlow to the new API defined in
+Corda 4. It utilises the flow versioning API to determine what version of the flow is running on the counterparty and
+respond accordingly. This process can be generalised to any flow upgrade procedure that may be required.
+
+ 1. Deploy the nodes in their initial configuration by running `./gradlew deployNodes`, then `build/nodes/runnodes`.
+ This creates a network of 3 nodes, all running version 1 of the contract. PartyA and PartyB are using the original
+ version of the workflows, while PartyC is using version 2.
+ 
+ 2. Check that all the nodes can transact with each other. This can be done manually from the shell, or by running
+ `./gradlew issueBetweenNodes`. This will issue an obligation between each pair of nodes in the network.
+ 
+ 3. Now shut down the nodes and upgrade all nodes to version 2 of the workflows CorDapp. The upgrade part of this can be
+ carried out by running `scripts/upgradeNodes.sh workflows v2-new-finality-flow PartyA PartyB`.
+ This simply copies the workflow jars (and any additional configuration) to the cordapps/ directory on the nodes specified.
+ 
+ 4. Restart the nodes using `build/nodes/runnodes`. All nodes should now be running V2 of the workflows CorDapp. The nodes
+ should still be able to transact with each other. New obligations can again be issued between all nodes using 
+ `./gradlew issueBetweenNodes`, and old obligations can be settled using `./gradlew settleAllObligations`. The shell can
+ also be used to experiment.
+ 
+Note: If this example is compiled against Corda 4.0, then a bug will prevent this scenario from executing successfully.
+The issue is that the new version of `FinalityFlow` in 4.0 has a `targetPlatformVersion` gate that prevents the old API
+from being called if the app has set `targetPlatformVersion = 4`. To fix this, change the V2 workflows app to use
+`targetPlatformVersion = 3` instead. This issue is fixed in Corda 4.1.
```

Comment:
OK - I have my answer here (so sounds like we have an interoperability break for this scenario using 4.0). Will need to highlight this in the release notes for CE 4.0 and mention it is fixed in C4.1